test/tester.go

```diff
@@ -0,0 +1,261 @@
+package test
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"regexp"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/skynetlabs/pinner/api"
+	"github.com/skynetlabs/pinner/database"
+	"gitlab.com/NebulousLabs/errors"
+	"gitlab.com/SkynetLabs/skyd/build"
+)
+
+var (
+	testPortalAddr = "http://127.0.0.1"
+	testPortalPort = "6000"
+
+	// dontFollowRedirectsCheckRedirectFn is a function that instructs http.Client
+	// to return with the last user response, instead of following a redirect.
+	dontFollowRedirectsCheckRedirectFn = func(req *http.Request, via []*http.Request) error {
+		return http.ErrUseLastResponse
+	}
+)
+
+type (
+	// Tester is a simple testing kit. It starts a testing instance of the
+	// service and provides simplified ways to call the handlers.
+	Tester struct {
+		Ctx             context.Context
+		DB              *database.DB
+		FollowRedirects bool
+		Logger          *logrus.Logger
+
+		cancel context.CancelFunc
+	}
+)
+
+// NewDatabase returns a new DB connection based on the passed parameters.
+func NewDatabase(ctx context.Context, dbName string) (*database.DB, error) {
+	return database.NewCustomDB(ctx, SanitizeName(dbName), DBTestCredentials(), NewDiscardLogger())
+}
+
+// NewTester creates and starts a new Tester service.
+// Use the Close method for a graceful shutdown.
+func NewTester(dbName string) (*Tester, error) {
+	ctx := context.Background()
+	logger := NewDiscardLogger()
+
+	// Connect to the database.
+	db, err := NewDatabase(ctx, dbName)
+	if err != nil {
+		return nil, errors.AddContext(err, "failed to connect to the DB")
+	}
+
+	ctxWithCancel, cancel := context.WithCancel(ctx)
+
+	// The server API encapsulates all the modules together.
+	server, err := api.New(db, logger)
+	if err != nil {
+		cancel()
+		return nil, errors.AddContext(err, "failed to build the API")
+	}
+
+	// Start the HTTP server in a goroutine and gracefully stop it once the
+	// cancel function is called and the context is closed.
+	srv := &http.Server{
+		Addr:    ":" + testPortalPort,
+		Handler: server,
+	}
+	go func() {
+		_ = srv.ListenAndServe()
+	}()
+	go func() {
+		select {
+		case <-ctxWithCancel.Done():
+			_ = srv.Shutdown(context.TODO())
+		}
+	}()
```

Comment:
f/u This is not very graceful because the goroutines might just be terminated when the test ends without making sure the http server actually shut down. One way to do that gracefully is something like this

```go
tester.externShutdownDone = make(chan error)
go func() {
  tester.externShutdownErr = srv.ListenAndServe()
  close(tester.externShutdownDone)
}()
```

and then 
```go
func (t *Tester) Close() error {
  if err := t.staticSRV.Shutdown(someCtx); err != nil {
    return err
  }
  <-tester.externShutdownDone
  return tester.externShutdownErr
}
```

Reply:
F/U: https://github.com/SkynetLabs/pinner/issues/4